tests/validator/validator.cpp

```diff
@@ -1065,3 +1063,726 @@ TEST(Validator, validMathCnElements)
+    u8->setName("testunit8");
+    u8->addUnit("newton", 1.0);
+    u8->addUnit("pascal", -1.0);
+    u8->addUnit("sievert", -1.0);
+    libcellml::UnitsPtr u9 = std::make_shared<libcellml::Units>();
+    u9->setName("testunit9");
+    u9->addUnit("second", 2.0);
+    u9->addUnit("radian", -4.0);
+    u9->addUnit("steradian", 2.0);
+
+    // u13 != u14: testing that the mismatch is reported correctly.
+    libcellml::UnitsPtr u13 = std::make_shared<libcellml::Units>();
+    u13->setName("testunit13");
+    u13->addUnit("testunit1", "kilo", 2.0);
+    u13->addUnit("testunit8", 2.0);
+    libcellml::UnitsPtr u14 = std::make_shared<libcellml::Units>();
+    u14->setName("testunit14");
+    u14->addUnit("testunit1", 1.0);
+    u14->addUnit("testunit9", 2.0);
+
+    v1->setUnits(u1);
+    v8->setUnits(u8);
+    v9->setUnits(u9);
+    v13->setUnits(u13);
+    v14->setUnits(u14);
+
+    m->addUnits(u1);
+    m->addUnits(u8);
+    m->addUnits(u9);
+    m->addUnits(u13);
+    m->addUnits(u14);
+
+    // This one is fine but complicated: newton/(sievert.pascal) = second^2.radian^3.steradian^-4.
+    libcellml::Variable::addEquivalence(v8, v9);
+
+    // Off by (metre)^1: testing nested unit equivalence.
+    libcellml::Variable::addEquivalence(v13, v14);
+
+    validator.validateModel(m);
+
+    EXPECT_EQ(size_t(2), validator.errorCount());
+    for (size_t i = 0; i < validator.errorCount(); ++i) {
+        EXPECT_EQ(expectedErrors.at(i), validator.error(i)->description());
+    }
+}
+
+TEST(Validator, unitEquivalenceExponentMultiplierPrefixExponent)
+{
+    // This test is intended to demonstrate that the effect of different multiplicator sources (prefix, multiplier term)
+    // does not affect the equivalence of the underlying base variables.
+    // TODO: when warnings are implemented then the multiplier testing can be reinstated.
+
+    libcellml::Validator validator;
+    libcellml::ModelPtr model = std::make_shared<libcellml::Model>();
+
+    libcellml::ComponentPtr comp1 = std::make_shared<libcellml::Component>();
+    libcellml::ComponentPtr comp2 = std::make_shared<libcellml::Component>();
+
+    libcellml::VariablePtr v1 = std::make_shared<libcellml::Variable>();
+    libcellml::VariablePtr v2 = std::make_shared<libcellml::Variable>();
+
+    v1->setName("v1");
+    v2->setName("v2");
+
+    // millimetres
+    libcellml::UnitsPtr u1 = std::make_shared<libcellml::Units>();
+    u1->setName("u1");
+    u1->addUnit("metre", "milli"); // standard, prefix.
+
+    // mm^3
+    libcellml::UnitsPtr u2 = std::make_shared<libcellml::Units>();
+    u2->setName("u2");
+    u2->addUnit("u1", 3.0); // standard, exponent.
+
+    // mm^6
+    libcellml::UnitsPtr u3 = std::make_shared<libcellml::Units>();
+    u3->setName("u3");
+    u3->addUnit("u2", 2.0); // standard, exponent.
+
+    // m^6
+    libcellml::UnitsPtr u4 = std::make_shared<libcellml::Units>();
+    u4->setName("u4");
+    u4->addUnit("u3", 15, 1.0); // standard, prefix, exponent.
```

Comment:
I am confused here... shouldn't it be `18` rather than `15`? Or is the idea just to have variables with equivalent units as opposed to exactly the same unit?

Reply:
In fact, the current validation code seems to be checking only for equivalent units (and rightly so when it comes to validation), which means that we could in fact use any prefix we want here.

Reply:
Hi @agarny Yes, apparently we don't mind about the multiplier/prefix mismatch of units across equivalent variables.  It will be raised as a "warning" level when these are in and I can finally get to working on the Issues structure for non-fatal stuff.

Reply:
Fair enough, however it might be worth making a small comment about this in the test. It's just that when I saw it, I got confused and couldn't help myself but working out the units equivalence by hand to make sure that it was "correct" (hence wastinga bit of time).